IN-381 CRUD category assignments

adessy opened the pull request:
### Implements those new endpoints 
...from DB tables to routes.rb

- GET inputs: GET /insights/views/:view_id/inputs
- GET input: GET /insights/views/:view_id/inputs/:id 
- Assign categories: POST /insights/views/:view_id/inputs/:id/categories
```json
{
    "data": [
        {"type":"category", "id": "category-uuid"},
        {"type":"category", "id": "category-uuid"}
    ]
}
```

- Dissmiss all suggestions: DELETE /insights/views/:view_id/inputs/:input_id/suggested_categories
- Dissmiss 1 suggestion: DELETE /insights/views/:view_id/inputs/:input_id/suggested_categories/:category_id
- Remove all category assignments: DELETE /insights/views/:view_id/inputs/:input_id/categories
- Remove 1 category assignments: DELETE /insights/views/:view_id/inputs/:input_id/categories/:category_id
- Delete all categories: DELETE /insights/views/:view_id/categories

### What's missing
(but was initially in the scope)

The processed status of the input:
```json
{
  "data": {
    "type": "input",
    "id": "idea-uuid",
    "attributes": {
      "processed": false <- HERE!
    },
    "relationship": {
      "source": {"data": {"id": "idea-uuid", "type": "idea"}},
      "categories": {
        "data": [
          {"type": "category", "id": "category-uuid"},
          {"type": "category", "id": "category-uuid"}
        ]
      },
      "suggested_categories": {
        "data": [
          {"type": "category", "id": "category-uuid"},
          {"type": "category", "id": "category-uuid"}
        ]
      }
    }
  },
  "included": ["..."]
}
```

This flag will be used to know if an input must appear in the inbox (_recently added_) or in the _REST_ category. There are two reasons why I decided to postpone this:
- wanted to keep this PR under 1K lines
- not exactly sure when this flag should change to `true`: explicit request or category assignment?

adessy committed:
IN-381 Introduce basic DB table for category assignments

adessy committed:
IN-381 Add basic implementation of CategoryAssignment model + factory

adessy committed:
IN-381 Simplify/fix the DB table for category assignments

adessy committed:
IN-381 Add basic routes for view inputs and category assignments

- GET input(s)
- POST assignments
- PATCH assignments
- DELETE assignments

adessy committed:
Merge branch 'e2e-flow/IN-379/crud-categories' into e2e-flow/IN-381/crud-category-assignments

adessy committed:
IN-381 Add `approved` column to category assignment

adessy committed:
IN-381 Test category assignments are approved by default

adessy committed:
IN-381 Add acceptance tests for view inputs

+ a few fixes

adessy committed:
IN-381 Move assignments to a dedicated controller

+ add an endpoint to DELETE a single assignment

adessy committed:
IN-381 Fix: adding categories should be a POST

adessy committed:
IN-381 Add routes to list & discard category suggestions

adessy committed:
IN-381 Convert suggestions into approved assignments

When adding new category assignments to an input, if the assignment matches a suggestion, we approve the suggestion which is thereby converted into an actual assignment.

adessy committed:
IN-381 Fix 'delete suggestions' endpoint

adessy committed:
IN-381 Rework category assignments controllers and service

adessy committed:
IN-381 Add acceptance specs for category assignments

adessy committed:
Merge branch 'master' into e2e-flow/IN-381/crud-category-assignments

adessy committed:
IN-381 Add acceptance specs for category suggestion & assignments

adessy committed:
IN-381 Add tests for input & category serializers

adessy committed:
IN-382 Add an InputsFinder to paginate and allow search

adessy committed:
IN-382 Add acceptance tests for pagination & text search of inputs

SaraBCL commented on back/db/migrate/20210601061247_create_category_assignments.insights.rb:
```diff
@@ -0,0 +1,16 @@
+    create_table :insights_category_assignments, id: :uuid do |t|
+      t.references :category, type: :uuid, null: false, index: true, foreign_key: { to_table: :insights_categories }
+      t.references :input, type: :uuid, null: false, index: true, polymorphic: true
+      t.boolean :approved, null: false, index: true, default: true
```
Just an idea : might be cool to do something about the confidence score in the future

adessy replied:
Where should we log this? in the spec file?

SaraBCL commented on back/engines/commercial/insights/app/controllers/insights/web_api/v1/category_assignments_controller.rb:
```diff
@@ -0,0 +1,71 @@
+        assignments = assignment_service.add_assignments(input, categories)
+        errors = assignments.map(&:errors).select(&:any?)
+        if errors.present?
+          # TODO: Improvement: report all errors, not only the first one.
```
Right, let's not forget this :)

adessy replied:
I don't feel tooo bad about this one :-) Do you know about similar cases in the code base?

SaraBCL replied:
No you're right, I don't think we're going to do anything super meaningful with that info in the front-end for now in any case :)

SaraBCL replied:
No you're right, I don't think we're going to do anything super meaningful with that info in the front-end for now in any case :)

adessy replied:
Resolving it for now then.

SaraBCL commented on back/db/migrate/20210601061247_create_category_assignments.insights.rb:
```diff
@@ -0,0 +1,16 @@
+      t.references :category, type: :uuid, null: false, index: true, foreign_key: { to_table: :insights_categories }
+      t.references :input, type: :uuid, null: false, index: true, polymorphic: true
+      t.boolean :approved, null: false, index: true, default: true
+
```
I thik it'd be really useful for the front-end to implement the search or cerate field in the modal to add a `accepts_nested_attributes_for :insights_category`.

adessy replied:
Not sure to understand how this would help. Do you have minimal example in mind?

I already started to implement the search in this [PR](https://github.com/CitizenLabDotCo/citizenlab/pull/302).

SaraBCL replied:
When you're looking at an input in the side modal, there's a field you can type in to add a category to an input, if no category was found with what you typed, this category can be created and assigned instead. Without the nested attributes, front-end will have to make two requests : create the category, await the result and then create the assignment.

adessy replied:
Ah I see! Gonna log this one to Jira not to block this PR.

adessy replied:
https://citizenlab.atlassian.net/browse/IN-405

SaraBCL replied:
Sounds good, thanks ! :)

adessy committed:
IN-382 Add tests for InputsFinder

SaraBCL commented on back/engines/commercial/insights/app/controllers/insights/web_api/v1/category_assignments_controller.rb:
```diff
@@ -0,0 +1,71 @@
+        @input ||= view.scope.ideas.find(params.require(:input_id))
+      end
+
+      # @return [Array<Insights::CategoryAssignment>]
```
```suggestion
      # @return [Array<Insights::Category>]
```

adessy replied:
Good catch!

SaraBCL reviewed (approved):
Neat. just a few comments and suggestions, but definitely not blocking.

SaraBCL commented on back/engines/commercial/insights/app/models/insights/category_assignment.rb:
```diff
@@ -0,0 +1,12 @@
+
+module Insights
+  class CategoryAssignment < ::ApplicationRecord
+    belongs_to :category
```
You're not using a dependency to delete assignments when a category is deleted ? How is it done, then ?

adessy replied:
Very good point! I realised that this morning. It's also true for the idea-assignment relationship. To be fixed. I already fixed it for ideas in the other PR (didn't want to push new modifications here while u were reviewing it).

SaraBCL commented on back/engines/commercial/insights/spec/models/insights/category_assignment_spec.rb:
```diff
@@ -0,0 +1,67 @@
+
+    specify { expect(assignment).to be_valid }
+
+    it 'is not valid without a view' do
```
I don't think it's necessary to have a view here. I think that test is actually is failing :)

adessy replied:
Other good catch. Leftover of a previous implementation where assignment also had references to views. But it was redundant with categories. Still have to check why the assignment is not valid though. Maybe it's the failing test actually 🤞

adessy replied:
Good news: the test was indeed failing, + a few other tests related to dependencies between categories/views/inputs and assignments. The test related to the assignments-input dependency is still failing in this branch for now.

adessy committed:
IN-382 Add association from Idea to CategoryAssignment

Using a concern bc the goal is to reuse this with other types of inputs (e.g. comments).

adessy committed:
IN-382 Use a polymorphic association for Input#insights_category_assignments

adessy committed:
IN-382 Add a category filter to GET /inputs

adessy committed:
Update back/engines/commercial/insights/app/controllers/insights/web_api/v1/category_assignments_controller.rb

IN-381 Fix type hinting

Co-authored-by: Sara <38424075+SaraBCL@users.noreply.github.com>

adessy committed:
IN-381 Add missing 'assignments' association to category

adessy committed:
Merge remote-tracking branch 'origin/e2e-flow/IN-381/crud-category-assignments' into e2e-flow/IN-381/crud-category-assignments

adessy committed:
Merge branch 'e2e-flow/IN-381/crud-category-assignments' into IN-382/view-inputs-pagination-filters

adessy committed:
IN-382 Add tests for the category filter of GET .../inputs

adessy committed:
Merge branch 'master' into e2e-flow/IN-381/crud-category-assignments

adessy committed:
Update back/engines/commercial/insights/spec/lib/finders/insights/inputs_finder_spec.rb

Co-authored-by: Sara <38424075+SaraBCL@users.noreply.github.com>

adessy committed:
IN-383 Fix category filter in input finder

adessy committed:
Merge remote-tracking branch 'origin/IN-382/view-inputs-pagination-filters' into IN-382/view-inputs-pagination-filters

adessy committed:
Merge pull request #302 from CitizenLabDotCo/IN-382/view-inputs-pagination-filters

adessy committed:
IN-381 Fix typos + rubocop
